### DISCUSSION TITLE
question about "ask-s" with "time-out"

### DISCUSSION THREAD

**Discussion by ijitai** [mdbergmann/cl-gserver#46] [General]
Suppose there are A1, A2, ..., A10, 10 actors, they ask-s B1, B2, ..., B10 with "time-out" in the same time.
And B1, B2, B10 are all time-out, and they throw 10 handler-error in the same time. How does the handler deal with this situation? Or ask-s should not throw exception, and just return a "fail" value?

**Comment by mdbergmann** [comment]
Hi.
I've turned you question into a 'discussion'.

I'm not exactly sure I understand the question.
An `ask-s` with timeout returns a result: `(cons :handler-error ask-timeout-condition)` if no result was retrieved from B. No exception is thrown, at least not to the caller of `ask-s`.

**Comment by ijitai** [comment]
I means if there are 10 handler-errors(time-out)  in the same time, how does the handler deal with it?

in Actor a1
(handler-case
 (ask-s B1 :time-out 1)
 (ask-timeout-condition 
   (do-sth.)))

in Actor a2
(handler-case
 (ask-s B2 :time-out 1)
 (ask-timeout-condition 
   (do-sth.)))

....
in Actor a10
(handler-case
 (ask-s B10 :time-out 1)
 (ask-timeout-condition 
   (do-sth.)))

if B1, B2, ... B10 are all time-out, and throw 10 errors in the same time,
 How do these handlers deal with it?
In multi-threads enviroment, one error/exception may not attach one fixed handler-case; it is not in single thread.
 A1's handler-case deal with B1's condition or B2's, B3's... A2's handler-case deal with B2's condtion or B1's, B2's or ... B10's.

**Reply by mdbergmann** [reply]
I still don't fully understand your question. Let me clarify...
`ask-s` does _not_ raise a condition in case it gets no response from the actor it asks.
`ask-s` provides a return value indicating the timeout-condition, so you could just do:

```
;; in A1
(let ((ask-result (ask-s B1 :time-out 1)))
  (unless (and (eq :handler-error (car ask-result)) (typep (cdr ask-result) ask-timeout))
    (do-something)))
```

The `ask-s` blocks until B sends a response or a timeout occurs.
It is guarantied that the result of `ask-s` is from the actor it asked.

**Comment by ijitai** [comment]
Thanks. I get it.

**Reply by mdbergmann** [reply]
One note on this.

As the readme indicates, you shouldn't do blocking operations in the :receive function, because it can cause dead-locks, except you know exactly what you're doing.
You can either use a separate thread for the `ask-s`, or use the async `ask` and do `on-complete` on the future.

**Comment by ijitai** [comment]
Thanks for your note.
